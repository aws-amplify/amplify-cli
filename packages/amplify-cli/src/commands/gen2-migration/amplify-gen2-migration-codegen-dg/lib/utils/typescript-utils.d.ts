export declare const generateImportStatement: (moduleName: string, importPath: string) => string;
export declare const generateExportStatement: (exportName: string) => string;
//# sourceMappingURL=typescript-utils.d.ts.map
