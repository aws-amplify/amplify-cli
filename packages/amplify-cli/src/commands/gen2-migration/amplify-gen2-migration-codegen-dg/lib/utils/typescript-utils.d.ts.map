{"version":3,"file":"typescript-utils.d.ts","sourceRoot":"","sources":["../../src/utils/typescript-utils.ts"],"names":[],"mappings":"AACA,eAAO,MAAM,uBAAuB,eAAgB,MAAM,cAAc,MAAM,KAAG,MAEhF,CAAC;AAEF,eAAO,MAAM,uBAAuB,eAAgB,MAAM,KAAG,MAE5D,CAAC"}