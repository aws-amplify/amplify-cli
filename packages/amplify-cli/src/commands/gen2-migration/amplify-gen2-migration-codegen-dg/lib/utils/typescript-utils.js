'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.generateExportStatement = exports.generateImportStatement = void 0;
const generateImportStatement = (moduleName, importPath) => {
  return `import { ${moduleName} } from '${importPath}';`;
};
exports.generateImportStatement = generateImportStatement;
const generateExportStatement = (exportName) => {
  return `export const ${exportName} = `;
};
exports.generateExportStatement = generateExportStatement;
//# sourceMappingURL=typescript-utils.js.map
