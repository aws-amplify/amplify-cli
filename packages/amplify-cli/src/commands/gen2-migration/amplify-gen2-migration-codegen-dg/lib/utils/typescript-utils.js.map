{"version":3,"file":"typescript-utils.js","sourceRoot":"","sources":["../../src/utils/typescript-utils.ts"],"names":[],"mappings":";;;AACO,MAAM,uBAAuB,GAAG,CAAC,UAAkB,EAAE,UAAkB,EAAU,EAAE;IACxF,OAAO,YAAY,UAAU,YAAY,UAAU,IAAI,CAAC;AAC1D,CAAC,CAAC;AAFW,QAAA,uBAAuB,2BAElC;AAEK,MAAM,uBAAuB,GAAG,CAAC,UAAkB,EAAU,EAAE;IACpE,OAAO,gBAAgB,UAAU,KAAK,CAAC;AACzC,CAAC,CAAC;AAFW,QAAA,uBAAuB,2BAElC"}