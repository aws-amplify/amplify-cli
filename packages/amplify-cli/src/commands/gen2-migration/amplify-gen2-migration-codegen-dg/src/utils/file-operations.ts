// File and directory operations utilities
import * as fs from 'fs/promises';
import * as path from 'path';

export const ensureDirectory = async (dirPath: string): Promise<void> => {
  await fs.mkdir(dirPath, { recursive: true });
};

export const writeFile = async (filePath: string, content: string): Promise<void> => {
  await ensureDirectory(path.dirname(filePath));
  await fs.writeFile(filePath, content);
};
