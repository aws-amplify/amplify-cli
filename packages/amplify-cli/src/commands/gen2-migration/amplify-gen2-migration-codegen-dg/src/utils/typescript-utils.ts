// TypeScript code generation utilities
export const generateImportStatement = (moduleName: string, importPath: string): string => {
  return `import { ${moduleName} } from '${importPath}';`;
};

export const generateExportStatement = (exportName: string): string => {
  return `export const ${exportName} = `;
};
