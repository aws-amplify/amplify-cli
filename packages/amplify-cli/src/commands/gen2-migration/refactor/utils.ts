export default function extractStackNameFromId(stackId: string): string {
  return stackId.split('/')[1];
}
