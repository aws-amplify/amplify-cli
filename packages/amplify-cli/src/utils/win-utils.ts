import * as fs from 'fs-extra';
import { command } from 'execa';
import { oldVersionPath, pendingDeletePath, tmpRegPath } from './win-constants';

const regPath = 'HKLM\\SYSTEM\\CurrentControlSet\\Control\\Session Manager';
const regKey = 'PendingFileRenameOperations';
const regPreamble = `Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\\SYSTEM\\CurrentControlSet\\Control\\Session Manager]
"${regKey}"=hex(7):`;

/**
 * On Windows when upgrading, the current binary can't be deleted, only renamed.
 * This function checks for the renamed binary (the old one after upgrading) and removes it on the next run of the CLI
 *
 */
export const deleteOldVersion = () => {
  if (process.platform.startsWith('win') && fs.existsSync(oldVersionPath)) {
    try {
      fs.removeSync(oldVersionPath);
    } catch (err) {
      console.warn(`Failed to clean up previous CLI installation at [${oldVersionPath}].`);
      console.log(err);
      console.warn('Make sure this file is not open anywhere else.');
    }
  }
};

/**
 * On Windows the current binary cannot uninstall itself
 * There is a registry value that can be set to mark files for deletion on the next reboot of the system
 * This function marks the Amplify binary for deletion
 */
export const setRegPendingDelete = async () => {
  if (!process.platform.startsWith('win')) return;
  let regQueryOutput = '';
  try {
    ({ stdout: regQueryOutput } = await command(`reg query "${regPath}" /v ${regKey}`, { shell: 'cmd.exe' }));
  } catch (err) {
    // intentionally swallow this
    // it means the regKey doesn't exist which is fine, it will be created
  }
  const startIdx = Math.max(0, regQueryOutput.indexOf('\\??'));
  const currentPaths = regQueryOutput
    .slice(startIdx)
    .trim()
    .split('\\0')
    .filter(p => !!p);
  const newPaths = currentPaths.concat(`\\??\\${pendingDeletePath}`);
  const newHex = newPaths
    .map(strToLittleEndianHex)
    .map(hexArr => hexArr.join(','))
    .join(',00,00,00,00,')
    .concat(',00,00,00,00,00,00');
  const regContent = `${regPreamble}${newHex}`;
  await fs.writeFile(tmpRegPath, regContent);
  await command(`reg import "${tmpRegPath}"`, { shell: 'cmd.exe' });
  await fs.remove(tmpRegPath);
};

const strToLittleEndianHex = (str: string) => {
  const hexArr: string[] = [];
  for (let i = 0; i < str.length; i++) {
    let hexCode = str.charCodeAt(i).toString(16);
    switch (hexCode.length) {
      case 1:
        hexArr.push(`0${hexCode}`);
        hexArr.push('00');
        break;
      case 2:
        hexArr.push(hexCode);
        hexArr.push('00');
        break;
      case 3:
        hexArr.push(hexCode.slice(1));
        hexArr.push(`0${hexCode[0]}`);
        break;
      case 4:
        hexArr.push(hexCode.slice(2));
        hexArr.push(hexCode.slice(0, 2));
        break;
      default:
        throw new Error(`Could not convert hex code ${hexCode} into 16 bit little endian format`);
    }
  }
  return hexArr;
};
