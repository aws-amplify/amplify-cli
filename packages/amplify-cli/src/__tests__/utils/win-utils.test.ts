import * as fs from 'fs-extra';
import { deleteOldVersion, setRegPendingDelete } from '../../utils/win-utils';
import execa from 'execa';

// save original process.platform
const originalPlatform = process.platform;

jest.mock('fs-extra');
const fs_mock = fs as jest.Mocked<typeof fs>;
fs_mock.existsSync.mockReturnValue(true);

jest.mock('amplify-cli-core', () => ({
  pathManager: {
    getHomeDotAmplifyDirPath: jest.fn().mockReturnValue('homedir/.amplify'),
  },
}));

jest.mock('../../utils/win-constants', () => ({
  oldVersionPath: 'homedir/.amplify/bin/amplify-old.exe',
  pendingDeletePath: 'homedir/pending-delete.exe',
  tmpRegPath: '/tmp/path',
}));

jest.mock('execa');
const execa_mock = execa as jest.Mocked<typeof execa>;

describe('delete old version', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    Object.defineProperty(process, 'platform', {
      value: 'win32',
    });
  });

  afterAll(() => {
    Object.defineProperty(process, 'platform', {
      value: originalPlatform,
    });
  });

  it('doesnt do anything if not on windows', () => {
    Object.defineProperty(process, 'platform', {
      value: 'darwin',
    });
    deleteOldVersion();
    expect(fs_mock.removeSync.mock.calls.length).toBe(0);
  });

  it('doesnt do anything if old version doesnt exist', () => {
    fs_mock.existsSync.mockReturnValueOnce(false);
    deleteOldVersion();
    expect(fs_mock.removeSync.mock.calls.length).toBe(0);
  });

  it('removes old version when present', () => {
    deleteOldVersion();
    expect(fs_mock.removeSync.mock.calls[0][0]).toMatchInlineSnapshot(`"homedir/.amplify/bin/amplify-old.exe"`);
  });

  it('prints warning if old version cannot be deleted', () => {
    const consoleWarnSpy = jest.spyOn(global.console, 'warn');
    fs_mock.removeSync.mockImplementationOnce(() => {
      throw new Error('test error removing old binary');
    });
    deleteOldVersion();
    expect(consoleWarnSpy.mock.calls.length).toBe(2);
    expect(consoleWarnSpy.mock.calls[0][0]).toMatchInlineSnapshot(
      `"Failed to clean up previous CLI installation at [homedir/.amplify/bin/amplify-old.exe]."`,
    );
    expect(consoleWarnSpy.mock.calls[1][0]).toMatchInlineSnapshot(`"Make sure this file is not open anywhere else."`);
  });
});

describe('set registry pending delete', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    Object.defineProperty(process, 'platform', {
      value: 'win32',
    });
  });

  afterAll(() => {
    Object.defineProperty(process, 'platform', {
      value: originalPlatform,
    });
  });

  it('doesnt do anything if not on windows', async () => {
    Object.defineProperty(process, 'platform', {
      value: 'darwin',
    });
    await setRegPendingDelete();
    expect(execa_mock.command.mock.calls.length).toBe(0);
  });

  it('handles regKey does not exist', async () => {
    execa_mock.command.mockImplementationOnce(() => {
      throw new Error('error when reg key does not exist');
    });
    await setRegPendingDelete();
    expect(fs_mock.writeFile.mock.calls[0][1]).toMatchSnapshot();
  });

  it('handles regKey empty', async () => {
    execa_mock.command.mockResolvedValueOnce({ stdout: '' } as any);
    await setRegPendingDelete();
    expect(fs_mock.writeFile.mock.calls[0][1]).toMatchSnapshot();
  });

  it('handles unicode characters in registry', async () => {
    execa_mock.command.mockResolvedValueOnce({ stdout: '\\??\\C:\\ĐĦŉ\\你好.txt\\0' } as any);
    await setRegPendingDelete();
    expect(fs_mock.writeFile.mock.calls[0][1]).toMatchSnapshot();
  });

  it('writes reg import file, imports and cleans up import file', async () => {
    let writtenFile = false;
    let importedFile = false;
    fs_mock.writeFile.mockImplementationOnce(() => {
      writtenFile = true;
    });
    execa_mock.command.mockResolvedValueOnce({ stdout: '' } as any).mockImplementationOnce((() => {
      if (!writtenFile) throw new Error('did not write to import file before importing!');
      importedFile = true;
    }) as any);
    fs_mock.remove.mockImplementationOnce(() => {
      if (!importedFile) throw new Error('removed import file before importing!');
    });

    await setRegPendingDelete();
    const tmpRegPath = '/tmp/path';
    expect(fs_mock.writeFile.mock.calls[0][0]).toEqual(tmpRegPath);
    expect(fs_mock.writeFile.mock.calls[0][1]).toMatchSnapshot();
    expect(execa_mock.command.mock.calls[1][0]).toEqual(`reg import "${tmpRegPath}"`);
    expect(fs_mock.remove.mock.calls[0][0]).toEqual(tmpRegPath);
  });
});
