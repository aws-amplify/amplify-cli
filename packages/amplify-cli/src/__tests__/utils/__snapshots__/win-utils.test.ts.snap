// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set registry pending delete handles regKey does not exist 1`] = `
"Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\\\\SYSTEM\\\\CurrentControlSet\\\\Control\\\\Session Manager]
\\"PendingFileRenameOperations\\"=hex(7):5c,00,3f,00,3f,00,5c,00,68,00,6f,00,6d,00,65,00,64,00,69,00,72,00,2f,00,70,00,65,00,6e,00,64,00,69,00,6e,00,67,00,2d,00,64,00,65,00,6c,00,65,00,74,00,65,00,2e,00,65,00,78,00,65,00,00,00,00,00,00,00"
`;

exports[`set registry pending delete handles regKey empty 1`] = `
"Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\\\\SYSTEM\\\\CurrentControlSet\\\\Control\\\\Session Manager]
\\"PendingFileRenameOperations\\"=hex(7):5c,00,3f,00,3f,00,5c,00,68,00,6f,00,6d,00,65,00,64,00,69,00,72,00,2f,00,70,00,65,00,6e,00,64,00,69,00,6e,00,67,00,2d,00,64,00,65,00,6c,00,65,00,74,00,65,00,2e,00,65,00,78,00,65,00,00,00,00,00,00,00"
`;

exports[`set registry pending delete handles unicode characters in registry 1`] = `
"Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\\\\SYSTEM\\\\CurrentControlSet\\\\Control\\\\Session Manager]
\\"PendingFileRenameOperations\\"=hex(7):5c,00,3f,00,3f,00,5c,00,43,00,3a,00,5c,00,10,01,26,01,49,01,5c,00,60,4f,7d,59,2e,00,74,00,78,00,74,00,00,00,00,00,5c,00,3f,00,3f,00,5c,00,68,00,6f,00,6d,00,65,00,64,00,69,00,72,00,2f,00,70,00,65,00,6e,00,64,00,69,00,6e,00,67,00,2d,00,64,00,65,00,6c,00,65,00,74,00,65,00,2e,00,65,00,78,00,65,00,00,00,00,00,00,00"
`;

exports[`set registry pending delete writes reg import file, imports and cleans up import file 1`] = `
"Windows Registry Editor Version 5.00

[HKEY_LOCAL_MACHINE\\\\SYSTEM\\\\CurrentControlSet\\\\Control\\\\Session Manager]
\\"PendingFileRenameOperations\\"=hex(7):5c,00,3f,00,3f,00,5c,00,68,00,6f,00,6d,00,65,00,64,00,69,00,72,00,2f,00,70,00,65,00,6e,00,64,00,69,00,6e,00,67,00,2d,00,64,00,65,00,6c,00,65,00,74,00,65,00,2e,00,65,00,78,00,65,00,00,00,00,00,00,00"
`;
